webflux spans are not collected if upstream connection is broken

```text
2023-05-08 22:06:51.747 ERROR 22627 --- [tor-http-nio-10] a.w.r.e.AbstractErrorWebExceptionHandler : [bTxId: 8a1b3c9110b64e45aebcb92a1b1655ff, bSpanId: 5863edb700000343, bMode: T] [70cb4fa0-209]  500 Server Error for HTTP POST "/api/user/register"

io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:29526
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	*__checkpoint ⇢ org.springframework.cloud.gateway.filter.WeightCalculatorWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ HTTP POST "/api/user/register" [ExceptionHandlingWebHandler]
Original Stack Trace:
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.pollConnect(Native Method) ~[na:na]
	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672) ~[na:na]
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:946) ~[na:na]
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:337) ~[netty-transport-4.1.86.Final.jar:4.1.86.Final]
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:334) ~[netty-transport-4.1.86.Final.jar:4.1.86.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:776) ~[netty-transport-4.1.86.Final.jar:4.1.86.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724) ~[netty-transport-4.1.86.Final.jar:4.1.86.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650) ~[netty-transport-4.1.86.Final.jar:4.1.86.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[netty-transport-4.1.86.Final.jar:4.1.86.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[netty-common-4.1.86.Final.jar:4.1.86.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.86.Final.jar:4.1.86.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.86.Final.jar:4.1.86.Final]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]

2023-05-08 22:06:51.747  WARN 22627 --- [tor-http-nio-10] o.b.a.o.t.context.impl.TracingContext    : [bTxId: 8a1b3c9110b64e45aebcb92a1b1655ff, bSpanId: 58637e5e00000342, bMode: T] TraceContext does not finish correctly. [4] spans are still remained in the stack. Please adding -Dbithon.tracing.debug=true parameter to your application to turn on the span life time message to debug. Remained spans: 
[TraceSpan[name=webflux, traceId=8a1b3c9110b64e45aebcb92a1b1655ff, spanId=5863b76200000340, parentId=null, clazz=org.springframework.http.server.reactive.ReactorHttpHandlerAdapter, method=apply, kind=SERVER, cost=-1683554811745019(micro seconds)], 
TraceSpan[name=filter, traceId=8a1b3c9110b64e45aebcb92a1b1655ff, spanId=5863a3f100000341, parentId=5863b76200000340, clazz=org.springframework.cloud.gateway.filter.AdaptCachedBodyGlobalFilter, method=filter, kind=NONE, cost=4(micro seconds)], 
TraceSpan[name=filter, traceId=8a1b3c9110b64e45aebcb92a1b1655ff, spanId=58637e5e00000342, parentId=5863b76200000340, clazz=org.springframework.cloud.gateway.filter.NettyRoutingFilter, method=filter, kind=NONE, cost=1417(micro seconds)], 
TraceSpan[name=httpclient, traceId=8a1b3c9110b64e45aebcb92a1b1655ff, spanId=5863edb700000343, parentId=58637e5e00000342, clazz=reactor.netty.http.client.HttpClientFinalizer, method=send, kind=CLIENT, cost=675(micro seconds)]]
```

--- a/agent/agent-observability/src/main/java/org/bithon/agent/observability/tracing/context/TraceContextListener.java
+++ b/agent/agent-observability/src/main/java/org/bithon/agent/observability/tracing/context/TraceContextListener.java
@@ -41,7 +41,7 @@ public class TraceContextListener {
     public TraceContextListener() {
         TraceConfig config = ConfigurationManager.getInstance().getConfig(TraceConfig.class);
         if (config != null && config.isDebug()) {
-            listeners.add(new SpanEventLogger());
+            listeners.add(new SpanEventDebugLogger());
         }
     }
 
@@ -67,7 +67,7 @@ public class TraceContextListener {
         void onSpanFinished(ITraceSpan span);
     }
 
-    private static class SpanEventLogger implements IListener {
+    private static class SpanEventDebugLogger implements IListener {
         private final ILogAdaptor log = LoggerFactory.getLogger(TraceContextListener.class);
 
         @Override

--- a/agent/agent-observability/src/main/java/org/bithon/agent/observability/tracing/context/impl/TracingContext.java
+++ b/agent/agent-observability/src/main/java/org/bithon/agent/observability/tracing/context/impl/TracingContext.java
@@ -16,7 +16,9 @@
 
 package org.bithon.agent.observability.tracing.context.impl;
 
+import org.bithon.agent.configuration.ConfigurationManager;
 import org.bithon.agent.observability.tracing.Tracer;
+import org.bithon.agent.observability.tracing.config.TraceConfig;
 import org.bithon.agent.observability.tracing.context.ITraceContext;
 import org.bithon.agent.observability.tracing.context.ITraceSpan;
 import org.bithon.agent.observability.tracing.context.TraceContextListener;
@@ -26,6 +28,7 @@ import org.bithon.agent.observability.tracing.id.ISpanIdGenerator;
 import org.bithon.agent.observability.tracing.reporter.ITraceReporter;
 import org.bithon.component.commons.logging.LoggerFactory;
 import org.bithon.component.commons.time.Clock;
+import org.bithon.component.commons.utils.StringUtils;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -42,6 +45,7 @@ public class TracingContext implements ITraceContext {
     private final Clock clock = new Clock();
     private final String traceId;
     private final ISpanIdGenerator spanIdGenerator;
+    private final boolean isDebugEnabled;
     private ITraceReporter reporter;
 
     private boolean finished = false;
@@ -50,6 +54,7 @@ public class TracingContext implements ITraceContext {
                           ISpanIdGenerator spanIdGenerator) {
         this.traceId = traceId;
         this.spanIdGenerator = spanIdGenerator;
+        this.isDebugEnabled = ConfigurationManager.getInstance().getConfig(TraceConfig.class).isDebug();
     }
 
     @Override
@@ -98,11 +103,23 @@ public class TracingContext implements ITraceContext {
     @Override
     public void finish() {
         if (!spanStack.isEmpty()) {
-            LoggerFactory.getLogger(TracingContext.class).warn("TraceContext does not finish correctly. "
-                                                                       + "[{}] spans are still remained in the stack. "
-                                                                       + "Please adding -Dbithon.tracing.debug=true parameter to your application to turn on the span life time message to debug. Remained spans: \n{}",
-                                                               spans.size(),
-                                                               spans);
+            if (isDebugEnabled) {
+                LoggerFactory.getLogger(TracingContext.class)
+                             .warn(StringUtils.format("TraceContext does not finish correctly. "
+                                                      + "[%d] spans are still remained unfinished. This IS a bug.\nRemained spans: \n%s",
+                                                      spanStack.size(),
+                                                      spanStack),
+                                   new RuntimeException("Bug Detected"));
+            } else {
+                LoggerFactory.getLogger(TracingContext.class).warn("TraceContext does not finish correctly. "
+                                                                   + "[{}] spans are still remained unfinished. This IS a bug.\n"
+                                                                   + "Please adding -Dbithon.tracing.debug=true parameter to your application to turn on the span life time message to debug. \nRemained spans: \n{}",
+                                                                   spanStack.size(),
+                                                                   spanStack);
+            }
+
+            this.spans.clear();
+            this.spanStack.clear();
             return;
         }
 
@@ -131,14 +148,36 @@ public class TracingContext implements ITraceContext {
     void onSpanFinished(TracingSpan span) {
 
         if (spanStack.isEmpty()) {
-            // TODO: internal error
             TraceContextListener.getInstance().onSpanFinished(span);
+
+            if (isDebugEnabled) {
+                LoggerFactory.getLogger(TracingContext.class)
+                             .warn(StringUtils.format("Try to finish a span which is not in the stack. This IS a bug.\nCurrent span: \n%s",
+                                                      span),
+                                   new RuntimeException("Bug Detected"));
+            } else {
+                LoggerFactory.getLogger(TracingContext.class).warn("Try to finish a span which is not in the stack. This IS a bug.\n"
+                                                                   + "Please adding -Dbithon.tracing.debug=true parameter to your application to turn on the span life time message to debug. \nCurrent span: \n{}",
+                                                                   span);
+            }
             return;
         }
 
         if (!spanStack.peek().equals(span)) {
-            // TODO: internal error
             TraceContextListener.getInstance().onSpanFinished(span);
+
+            if (isDebugEnabled) {
+                LoggerFactory.getLogger(TracingContext.class)
+                             .warn(StringUtils.format("Try to finish a span which does not match the span in the stack. This IS a bug.\nCurrent span: \n%s, \n Unfinished Spans:\n%s",
+                                                      span,
+                                                      spanStack),
+                                   new RuntimeException("Bug Detected"));
+            } else {
+                LoggerFactory.getLogger(TracingContext.class).warn("Try to finish a span which is not in the stack. This IS a bug.\n"
+                                                                   + "Please adding -Dbithon.tracing.debug=true parameter to your application to turn on the span life time message to debug. \nCurrent span: \n{}, Unfinished Spans:\n{}",
+                                                                   span,
+                                                                   spanStack);
+            }
             return;
         }
 

--- a/agent/agent-observability/src/main/java/org/bithon/agent/observability/tracing/context/impl/TracingContext.java
+++ b/agent/agent-observability/src/main/java/org/bithon/agent/observability/tracing/context/impl/TracingContext.java
@@ -40,12 +40,13 @@ import java.util.Stack;
  */
 public class TracingContext implements ITraceContext {
 
+    private static final boolean IS_DEBUG_ENABLED = ConfigurationManager.getInstance().getConfig(TraceConfig.class).isDebug();
+
     private final Stack<ITraceSpan> spanStack = new Stack<>();
     private final List<ITraceSpan> spans = new ArrayList<>();
     private final Clock clock = new Clock();
     private final String traceId;
     private final ISpanIdGenerator spanIdGenerator;
-    private final boolean isDebugEnabled;
     private ITraceReporter reporter;
 
     private boolean finished = false;
@@ -54,7 +55,6 @@ public class TracingContext implements ITraceContext {
                           ISpanIdGenerator spanIdGenerator) {
         this.traceId = traceId;
         this.spanIdGenerator = spanIdGenerator;
-        this.isDebugEnabled = ConfigurationManager.getInstance().getConfig(TraceConfig.class).isDebug();
     }
 
     @Override
@@ -103,7 +103,7 @@ public class TracingContext implements ITraceContext {
     @Override
     public void finish() {
         if (!spanStack.isEmpty()) {
-            if (isDebugEnabled) {
+            if (IS_DEBUG_ENABLED) {
                 LoggerFactory.getLogger(TracingContext.class)
                              .warn(StringUtils.format("TraceContext does not finish correctly. "
                                                       + "[%d] spans are still remained unfinished. This IS a bug.\nRemained spans: \n%s",
@@ -150,7 +150,7 @@ public class TracingContext implements ITraceContext {
         if (spanStack.isEmpty()) {
             TraceContextListener.getInstance().onSpanFinished(span);
 
-            if (isDebugEnabled) {
+            if (IS_DEBUG_ENABLED) {
                 LoggerFactory.getLogger(TracingContext.class)
                              .warn(StringUtils.format("Try to finish a span which is not in the stack. This IS a bug.\nCurrent span: \n%s",
                                                       span),
@@ -166,7 +166,7 @@ public class TracingContext implements ITraceContext {
         if (!spanStack.peek().equals(span)) {
             TraceContextListener.getInstance().onSpanFinished(span);
 
-            if (isDebugEnabled) {
+            if (IS_DEBUG_ENABLED) {
                 LoggerFactory.getLogger(TracingContext.class)
                              .warn(StringUtils.format("Try to finish a span which does not match the span in the stack. This IS a bug.\nCurrent span: \n%s, \n Unfinished Spans:\n%s",
                                                       span,

--- a/agent/agent-observability/src/main/java/org/bithon/agent/observability/tracing/context/impl/TracingSpan.java
+++ b/agent/agent-observability/src/main/java/org/bithon/agent/observability/tracing/context/impl/TracingSpan.java
@@ -181,9 +181,10 @@ class TracingSpan implements ITraceSpan {
     @Override
     public void finish() {
         if (this.endTime != 0) {
-
+            // This span has already been closed, this is a bug
             return;
         }
+
         this.endTime = context().clock().currentMicroseconds();
         try {
             this.tracingContext.onSpanFinished(this);

--- a/agent/agent-observability/src/main/java/org/bithon/agent/observability/tracing/context/impl/TracingSpan.java
+++ b/agent/agent-observability/src/main/java/org/bithon/agent/observability/tracing/context/impl/TracingSpan.java
@@ -181,6 +181,7 @@ class TracingSpan implements ITraceSpan {
     @Override
     public void finish() {
         if (this.endTime != 0) {
+
             return;
         }
         this.endTime = context().clock().currentMicroseconds();
@@ -201,6 +202,7 @@ class TracingSpan implements ITraceSpan {
                 ", method=" + this.method +
                 ", kind=" + this.kind +
                 ", cost=" + (this.endTime - this.startTime) + "(micro seconds)" +
+                ", exception=" + this.tags.getOrDefault(Tags.Exception.TYPE, "") +
                 "]";
     }
 }

--- a/agent/agent-plugins/spring-webflux/src/main/java/org/bithon/agent/plugin/spring/webflux/interceptor/AroundGatewayFilter$Filter.java
+++ b/agent/agent-plugins/spring-webflux/src/main/java/org/bithon/agent/plugin/spring/webflux/interceptor/AroundGatewayFilter$Filter.java
@@ -88,6 +88,7 @@ public class AroundGatewayFilter$Filter extends AroundInterceptor {
                                               .method(aopContext.getTargetClass(), aopContext.getMethod())
                                               .start());
 
+
         return InterceptionDecision.CONTINUE;
     }
 
@@ -111,8 +112,16 @@ public class AroundGatewayFilter$Filter extends AroundInterceptor {
             span.tag(aopContext.getException()).finish();
         } else {
             Mono<Void> originalReturning = aopContext.getReturningAs();
-            Mono<Void> replacedReturning = originalReturning.doOnError(span::tag)
-                                                            .doFinally((s) -> span.finish());
+            Mono<Void> replacedReturning = originalReturning.doOnError((throwable -> {
+                                                                // NOTE: DO NOT Replace the call by lambda which is suggested by the IDE,
+                                                                // because it's hard to debug the statement
+                                                                span.tag(throwable).finish();
+                                                            }))
+                                                            .doOnSuccess((s) -> {
+                                                                // NOTE: DO NOT Replace the call by lambda which is suggested by the IDE,
+                                                                // because it's hard to debug the statement
+                                                                span.finish();
+                                                            });
             aopContext.setReturning(replacedReturning);
         }
     }

--- a/agent/agent-plugins/spring-webflux/src/main/java/org/bithon/agent/plugin/spring/webflux/interceptor/HttpClientFinalizer$ResponseConnection.java
+++ b/agent/agent-plugins/spring-webflux/src/main/java/org/bithon/agent/plugin/spring/webflux/interceptor/HttpClientFinalizer$ResponseConnection.java
@@ -34,7 +34,7 @@ import reactor.netty.http.client.HttpClientResponse;
 import java.util.function.BiFunction;
 
 /**
- * see reactor.netty.http.client.HttpClientFinalizer#responseConnection
+ * see {@link reactor.netty.http.client.HttpClientFinalizer#responseConnection}
  *
  * @author frank.chen021@outlook.com
  * @date 27/11/21 1:57 pm

--- a/agent/agent-plugins/spring-webflux/src/main/java/org/bithon/agent/plugin/spring/webflux/interceptor/HttpClientFinalizer$Send.java
+++ b/agent/agent-plugins/spring-webflux/src/main/java/org/bithon/agent/plugin/spring/webflux/interceptor/HttpClientFinalizer$Send.java
@@ -33,7 +33,7 @@ import reactor.netty.http.client.HttpClientRequest;
 import java.util.function.BiFunction;
 
 /**
- * see reactor.netty.http.client.HttpClientFinalizer#send
+ * See {@link reactor.netty.http.client.HttpClientFinalizer#send}
  *
  * @author frank.chen021@outlook.com
  * @date 27/11/21 1:57 pm

--- a/agent/agent-plugins/spring-webflux/src/main/java/org/bithon/agent/plugin/spring/webflux/interceptor/ReactorHttpHandlerAdapter$Apply.java
+++ b/agent/agent-plugins/spring-webflux/src/main/java/org/bithon/agent/plugin/spring/webflux/interceptor/ReactorHttpHandlerAdapter$Apply.java
@@ -27,6 +27,7 @@ import org.bithon.agent.observability.metric.domain.web.HttpIncomingMetricsRegis
 import org.bithon.agent.observability.tracing.Tracer;
 import org.bithon.agent.observability.tracing.config.TraceConfig;
 import org.bithon.agent.observability.tracing.context.ITraceContext;
+import org.bithon.agent.observability.tracing.context.ITraceSpan;
 import org.bithon.agent.observability.tracing.context.TraceContextHolder;
 import org.bithon.agent.observability.tracing.context.TraceMode;
 import org.bithon.agent.observability.tracing.context.propagation.ITracePropagator;
@@ -89,7 +90,7 @@ public class ReactorHttpHandlerAdapter$Apply extends AroundInterceptor {
         TraceContextHolder.remove();
 
         //
-        // HttpServerRequest has an implementation of HttpServerOperation
+        // HttpServerRequest has an implementation of HttpServerOperation,
         // which is instrumented as IBithonObject
         //
         if (request instanceof IBithonObject) {
@@ -110,7 +111,11 @@ public class ReactorHttpHandlerAdapter$Apply extends AroundInterceptor {
                                 .configIfTrue(!traceConfig.getHeaders().getRequest().isEmpty(),
                                               (span) -> traceConfig.getHeaders()
                                                                    .getRequest()
-                                                                   .forEach((header) -> span.tag(Tags.Http.REQUEST_HEADER_PREFIX + header.toLowerCase(Locale.ENGLISH), request.requestHeaders().get(header))))
+                                                                   .forEach((header) -> span.tag(Tags.Http.REQUEST_HEADER_PREFIX
+                                                                                                 + header.toLowerCase(
+                                                                                                     Locale.ENGLISH),
+                                                                                                 request.requestHeaders()
+                                                                                                        .get(header))))
                                 .method(aopContext.getTargetClass(), aopContext.getMethod())
                                 .kind(SpanKind.SERVER)
                                 .start();
@@ -191,54 +196,65 @@ public class ReactorHttpHandlerAdapter$Apply extends AroundInterceptor {
             return;
         }
 
-        traceContext.currentSpan()
-                    .tag(Tags.Http.STATUS, String.valueOf(response.status().code()))
-                    .tag(t)
-                    .config((span -> {
-                        // extract headers in the response to tag
-                        if (!CollectionUtils.isEmpty(responseConfigs.getHeaders())) {
-                            HttpHeaders httpHeaders = response.responseHeaders();
-                            for (Map.Entry<String, String> entry : responseConfigs.getHeaders().entrySet()) {
-                                String headerName = entry.getKey();
-                                String tagName = entry.getValue();
-                                String tagValue = httpHeaders.get(headerName);
-                                if (tagValue != null) {
-                                    span.tag(tagName, tagValue);
-                                }
-                            }
-                        }
+        ITraceSpan span = traceContext.currentSpan();
+        if (span == null) {
+            // THIS IS a BUG. The unmatched 'finish' call already prints the error message, so we just return here
+            return;
+        }
 
-                        //
-                        // handle X-FORWARDED-FOR
-                        //
-                        if (xforwardTagName == null) {
-                            return;
-                        }
-                        InetSocketAddress remoteAddr = request.remoteAddress();
-                        if (remoteAddr == null) {
-                            return;
-                        }
+        if (!"webflux".equals(span.component())) {
+            // Directly close the context so that the tracing context prints the debug message
+            traceContext.finish();
+            return;
+        }
 
-                        String remoteAddrText = remoteAddr.getAddress().getHostAddress();
-                        List<String> xforwarded = request.requestHeaders().getAll(X_FORWARDED_FOR);
-                        if (xforwarded == null) {
-                            span.tag(xforwardTagName, remoteAddrText);
-                            return;
+        span.tag(Tags.Http.STATUS, String.valueOf(response.status().code()))
+            .tag(t)
+            .config((s -> {
+                // extract headers in the response to tag
+                if (!CollectionUtils.isEmpty(responseConfigs.getHeaders())) {
+                    HttpHeaders httpHeaders = response.responseHeaders();
+                    for (Map.Entry<String, String> entry : responseConfigs.getHeaders().entrySet()) {
+                        String headerName = entry.getKey();
+                        String tagName = entry.getValue();
+                        String tagValue = httpHeaders.get(headerName);
+                        if (tagValue != null) {
+                            s.tag(tagName, tagValue);
                         }
+                    }
+                }
 
-                        if (!xforwarded.contains(remoteAddrText)) { // prevent duplicates
-                            // xforwarded maybe unmodifiable, create a new container
-                            xforwarded = new ArrayList<>(xforwarded);
+                //
+                // handle X-FORWARDED-FOR
+                //
+                if (xforwardTagName == null) {
+                    return;
+                }
+                InetSocketAddress remoteAddr = request.remoteAddress();
+                if (remoteAddr == null) {
+                    return;
+                }
 
-                            xforwarded.add(remoteAddrText);
-                        }
+                String remoteAddrText = remoteAddr.getAddress().getHostAddress();
+                List<String> xforwarded = request.requestHeaders().getAll(X_FORWARDED_FOR);
+                if (xforwarded == null) {
+                    span.tag(xforwardTagName, remoteAddrText);
+                    return;
+                }
 
-                        // using join instead of using ObjectMapper
-                        // because in other modules such as tomcat-plugin, we directly get the header value and record it in the log
-                        // the value in that header is comma delimited
-                        span.tag(xforwardTagName, String.join(",", xforwarded));
-                    }))
-                    .finish();
+                if (!xforwarded.contains(remoteAddrText)) { // prevent duplicates
+                    // xforwarded maybe unmodifiable, create a new container
+                    xforwarded = new ArrayList<>(xforwarded);
+
+                    xforwarded.add(remoteAddrText);
+                }
+
+                // using join instead of using ObjectMapper
+                // because in other modules such as tomcat-plugin, we directly get the header value and record it in the log
+                // the value in that header is comma delimited
+                s.tag(xforwardTagName, String.join(",", xforwarded));
+            }))
+            .finish();
         traceContext.finish();
     }
 }

